libtenzir/builtins/operators/load_balance.cpp

```diff
@@ -0,0 +1,396 @@
+        .then(
+          [&](table_slice slice) {
+            result = std::move(slice);
+            ctrl.set_waiting(false);
+          },
+          [&](const caf::error& err) {
+            // This should never happen.
+            diagnostic::error("load balancer read failed: {}", err)
+              .emit(ctrl.diagnostics());
+          });
+      co_yield {};
+      // We signal completion with an empty table slice.
+      if (result.rows() == 0) {
+        TENZIR_WARN("load_balance_source detected end");
+        break;
+      }
+      TENZIR_WARN("load_balance_source read {} events", result.rows());
+      co_yield std::move(result);
+    }
+  }
+
+  auto optimize(const expression& filter, event_order order) const
+    -> optimize_result override {
+    TENZIR_UNUSED(filter, order);
+    return do_not_optimize(*this);
+  }
+
+  friend auto inspect(auto& f, load_balance_source& x) -> bool {
+    return f.apply(x.load_balancer_);
+  }
+
+private:
+  load_balancer_actor load_balancer_;
+};
+
+/// Inserts the result of a function call if the key does not exist yet.
+template <class Map, class Key, class F>
+auto get_or_compute(Map& map, Key&& key, F&& f) -> decltype(auto) {
+  class implicit_caller {
+  public:
+    explicit implicit_caller(F&& f) : f_{std::forward<F>(f)} {
+    }
+
+    explicit(false) operator typename Map::mapped_type() {
+      return std::forward<F>(f_)();
+    }
+
+  private:
+    F&& f_;
+  };
+  return map
+    .try_emplace(std::forward<Key>(key), implicit_caller(std::forward<F>(f)))
+    .first->second;
+}
+
+auto make_load_balancer(
+  load_balancer_actor::stateful_pointer<load_balancer_state> self,
+  std::vector<pipeline> pipes, shared_diagnostic_handler diagnostics,
+  metrics_receiver_actor metrics, uint64_t operator_index, bool is_hidden,
+  const node_actor& node) -> load_balancer_actor::behavior_type {
+  TENZIR_WARN("spawning load balancer");
+  self->attach_functor([] {
+    TENZIR_WARN("destroyed load balancer");
+  });
+  self->state.self = self;
+  self->state.diagnostics = std::move(diagnostics);
+  self->state.metrics = std::move(metrics);
+  self->state.operator_index = operator_index;
+  self->state.executors.reserve(pipes.size());
+  for (auto& pipe : pipes) {
+    pipe.prepend(std::make_unique<load_balance_source>(self));
+    auto has_terminal = false;
+    TENZIR_WARN("spawning inner executor");
+    auto executor = self->spawn<caf::monitored>(
+      pipeline_executor, pipe, self, self, node, has_terminal, is_hidden);
+    executor->attach_functor([] {
+      TENZIR_WARN("inner executor terminated");
+    });
+    self->request(executor, caf::infinite, atom::start_v)
+      .then(
+        []() {
+          TENZIR_WARN("started inner pipeline successfully");
+        },
+        [self](const caf::error& err) {
+          // This error should be enough to cause the outer pipeline to get
+          // cleaned up.
+          diagnostic::error(err).emit(self->state.diagnostics);
```

Comment:
Should we add a note here saying which configuration this was for?

Reply:
With the current implementation, that information is sadly not available here.

Reply:
Then let's not bother with it for now.